linux/src/interlocked_hl_linux.c

```diff
@@ -0,0 +1,281 @@
+                if (expected_operation_result > Ceiling)
+                {
+                    result = INTERLOCKED_HL_ERROR;
+                    break;
+                }
+                else
+                {
+                    /*Codes_SRS_INTERLOCKED_HL_02_005: [ Otherwise, InterlockedHL_Add64WithCeiling shall atomically write in Addend the sum of Addend and Value, succeed and return INTERLOCKED_HL_OK. ]*/
+                    if (interlocked_compare_exchange_64(Addend, expected_operation_result, addend_copy) == addend_copy)
+                    {
+                        /*Codes_SRS_INTERLOCKED_HL_02_007: [ In all failure cases InterlockedHL_Add64WithCeiling shall not modify Addend or originalAddend*/
+                        *originalAddend = addend_copy;
+                        result = INTERLOCKED_HL_OK;
+                        break;
+                    }
+                    else
+                    {
+                        /*go back to while(1)*/
+                    }
+                }
+            }
+        }
+    }
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, INTERLOCKED_HL_RESULT, InterlockedHL_WaitForValue, int32_t volatile_atomic*, address, int32_t, value, uint32_t, milliseconds)
+{
+    INTERLOCKED_HL_RESULT result;
+
+    /* Codes_SRS_INTERLOCKED_HL_01_002: [ If address is NULL, InterlockedHL_WaitForValue shall fail and return INTERLOCKED_HL_ERROR. ]*/
+    if (address == NULL)
+    {
+        result = INTERLOCKED_HL_ERROR;
+    }
+    else
+    {
+        int32_t current_value;
+
+        do
+        {
+            /* Codes_SRS_INTERLOCKED_HL_01_007: [ When interlocked_compare_exchange_64nAddress succeeds, the value at address shall be compared to the target value passed in value by using interlocked_add. ]*/
+            current_value = interlocked_add(address, 0);
+            if (current_value == value)
+            {
+                /* Codes_SRS_INTERLOCKED_HL_01_003: [ If the value at address is equal to value, InterlockedHL_WaitForValue shall return INTERLOCKED_HL_OK. ]*/
+                result = INTERLOCKED_HL_OK;
+                break;
+            }
+
+            /* Codes_SRS_INTERLOCKED_HL_01_008: [ If the value at address does not match, InterlockedHL_WaitForValue shall issue another call to WaitOnAddress. ]*/
+
+            /* Codes_SRS_INTERLOCKED_HL_01_004: [ If the value at address is not equal to value, InterlockedHL_WaitForValue shall wait until the value at address changes in order to compare it again to value by using WaitOnAddress. ]*/
+            /* Codes_SRS_INTERLOCKED_HL_01_005: [ When waiting for the value at address to change, the milliseconds argument value shall be used as timeout. ]*/
+
+            struct timespec timeout = {.tv_sec = milliseconds/1000, .tv_nsec = (milliseconds % 1000) * 1e6};
+            if (!futex(address, FUTEX_WAIT, value, &timeout, NULL, 0))
```

Comment:
I think we should just add a PAL for waitonaddress/wake and futex then this can all just go in common. The background here, is that we use InterlockedHL_WaitForAddress in many places in the code, but then we have calls directly to WakeByAddressSingle/WakeByAddressAll sprinkled throughout the code. We did want to move more of those calls to wake into InterlockedHL at some point, but I think that a good start may be to have a PAL for that call.

So I think we might just want to remove this linux implementation for now and then make this module generic.

Reply:
Adding the PAL for the wake/wait stuff should be a separate task